File: 29Items-Projects/7Algorithms/1-Sorting-and-Searching/src/python/algorithms/sorting/heap_sort.py
def heap_sort(values: list[int]) -> list[int]:
    result = values[:]
    n = len(result)

    for i in range(n // 2 - 1, -1, -1):
        _heapify(result, n, i)

    for end in range(n - 1, 0, -1):
        result[0], result[end] = result[end], result[0]
        _heapify(result, end, 0)

    return result


def _heapify(values: list[int], size: int, root: int) -> None:
    largest = root
    left = 2 * root + 1
    right = 2 * root + 2

    if left < size and values[left] > values[largest]:
        largest = left
    if right < size and values[right] > values[largest]:
        largest = right

    if largest != root:
        values[root], values[largest] = values[largest], values[root]
        _heapify(values, size, largest)


if __name__ == "__main__":
    sample = [42, 7, 19, 3, 7, 99, 1]
    print("HeapSort")
    print(f"input : {sample}")
    print(f"output: {heap_sort(sample)}")

File: 29Items-Projects/7Algorithms/1-Sorting-and-Searching/src/java/com/sortingsearching/sorting/HeapSort.java
package com.sortingsearching.sorting;

import java.util.Arrays;

public final class HeapSort {
    private HeapSort() {
    }

    public static int[] sort(int[] input) {
        int[] values = Arrays.copyOf(input, input.length);
        int n = values.length;

        for (int i = n / 2 - 1; i >= 0; i--) {
            heapify(values, n, i);
        }
        for (int end = n - 1; end > 0; end--) {
            swap(values, 0, end);
            heapify(values, end, 0);
        }
        return values;
    }

    private static void heapify(int[] values, int size, int root) {
        int largest = root;
        int left = 2 * root + 1;
        int right = 2 * root + 2;

        if (left < size && values[left] > values[largest]) {
            largest = left;
        }
        if (right < size && values[right] > values[largest]) {
            largest = right;
        }
        if (largest != root) {
            swap(values, root, largest);
            heapify(values, size, largest);
        }
    }

    private static void swap(int[] values, int i, int j) {
        int temp = values[i];
        values[i] = values[j];
        values[j] = temp;
    }

    public static void main(String[] args) {
        int[] sample = {42, 7, 19, 3, 7, 99, 1};
        System.out.println("HeapSort");
        System.out.println("input : " + Arrays.toString(sample));
        System.out.println("output: " + Arrays.toString(sort(sample)));
    }
}
